### PR TITLE
Only process traffic impacted by network policies

### PR DESCRIPTION
**Pull request by aojea**
optimize the datapath not having to send all packets to user space, only the ones that are subject of network policies

Fixes: https://github.com/kubernetes-sigs/kube-network-policies/issues/10, #31, #12

### CODE REVIEW

**Comment by aojea** [comment]
/assign @danwinship 

Dan please take a look

File: pkg/networkpolicy/networkpolicy.go [hunk 1]
```diff
@@ -10,6 +10,29 @@ import (
 	"k8s.io/klog/v2"
 )
 
+// getPodsForNetworkPolicyNode obtains all the Pods selected by the network policy for current Node
+func (c *Controller) getLocalPodsForNetworkPolicy(networkPolicy *networkingv1.NetworkPolicy) []*v1.Pod {
+	if networkPolicy == nil {
+		return nil
+	}
+	podSelector, err := metav1.LabelSelectorAsSelector(&networkPolicy.Spec.PodSelector)
```
Comment (danwinship) [inline pkg/networkpolicy/networkpolicy.go:18]: So FTR, with the existing code, if you have Pod A on Node 1 that denies all ingress, and Pod B on Node 2 tries to send to it, then Node 2 will drop the packets. With this new code, Node 2 would instead forward the packet to Node 1, and then Node 1 would drop it.

File: cmd/main.go [hunk 2]
```diff
@@ -64,11 +64,17 @@ func main() {
 		klog.Fatalf("error parsing metrics bind address %s : %v", metricsBindAddress, err)
 	}
 
+	nodeName := os.Getenv("MY_NODE_NAME")
+	if nodeName == "" {
+		klog.Fatalf("node name not set, please set the environment variable using the Downward API")
```
Comment (danwinship) [inline cmd/main.go:69]: It would be better to make this a command-line option, and for it to be optional.

**Review by danwinship — COMMENTED** [review]
This looks like it will work (though you lose all the optimization once you add even a single ANP)...

Another possibility would be to just add "local detector" options like kube-proxy (`--cluster-cidr`, use node `podCIDRs`, etc)

File: pkg/networkpolicy/networkpolicy.go (hunk 1 above)
Comment (aojea) [reply pkg/networkpolicy/networkpolicy.go:18]: yeah, is the trade off, but at least the rules are only evaluated once in one node, previously allowed rules were evaluated twice

File: cmd/main.go (hunk 2 above)
Comment (aojea) [reply cmd/main.go:69]: but now the reconciliation logic depends on the nodename and the installation yaml defines the downward API variable ... is there any case you see people may want to run without this ?

**Comment by aojea** [comment]
> This looks like it will work (though you lose all the optimization once you add even a single ANP)...

Cluster Wide Policies are hard to implement in the dataplane

> Another possibility would be to just add "local detector" options like kube-proxy (`--cluster-cidr`, use node `podCIDRs`, etc)

Since the controller already has the pod information this looks simpler, kube-proxy does not watch pods and need to depend on those heuristics

**Comment by danwinship** [comment]
It would still be better for it to be a command-line option, even if it's required. Required environment variables are terrible.

**Comment by aojea** [comment]
> It would still be better for it to be a command-line option, even if it's required. Required environment variables are terrible.

I just want to fix the flake, let me open an issue to not forget and do the change

**Pull request merged**